### PR TITLE
panic! -> abort

### LINKED ISSUE #79: LTO causes undefined references to core::panicking::panic
### Update

After #80, this only occurs with profile.dev (+debug-assertions) + LTO. To fix this we'll have to create newtypes over primitives types that `intrinsics::abort()` on bad inputs (division by zero) and overflow.

---

Getting a funky error when attempting to use `rustc-builtins` for nintendo 3DS shenanigans instead the prebuilt copy of `libcompiler-rt` that we've been using up to this point:

```
   Compiling ctru-rs v0.4.0 (https://github.com/FenrirWolf/ctru-rs?branch=builtins#27eb3caf)
   Compiling rustc_builtins v0.1.0 (https://github.com/japaric/rustc-builtins.git#69e93de9)
   Compiling alloc_system3ds v0.1.0 (https://github.com/rust3ds/alloc_system3ds?rev=da38c94#da38c941)
   Compiling ctru-sys v0.2.0 (https://github.com/FenrirWolf/ctru-rs?branch=builtins#27eb3caf)
   Compiling rust3ds-template v0.1.0 (file:///home/fenrir/projects/rust3ds-template)
error: linking with `arm-none-eabi-gcc` failed: exit code: 1
  |
  = note: "arm-none-eabi-gcc" "-specs=3dsx.specs" "-march=armv6k" "-mtune=mpcore" "-mfloat-abi=hard" "-mtp=soft" "-L" "/home/fenrir/.xargo/lib/rustlib/3ds/lib" "/home/fenrir/projects/rust3ds-template/target/3ds/release/rust3ds_template.0.o" "-o" "/home/fenrir/projects/rust3ds-template/target/3ds/release/rust3ds_template.elf" "-Wl,--gc-sections" "-nodefaultlibs" "-L" "/home/fenrir/projects/rust3ds-template/target/3ds/release/deps" "-L" "/opt/devkitPro/libctru/lib" "-L" "/home/fenrir/.xargo/lib/rustlib/3ds/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/tmp/rustc.n3iL9DJW0uUT/libctru-56422efc011c55b9.rlib" "/tmp/rustc.n3iL9DJW0uUT/librustc_builtins-5c8a87f356eb958c.rlib" "-lc" "-lm" "-lsysbase" "-lc" "-lsysbase" "-lc"
  = note: /tmp/rustc.n3iL9DJW0uUT/librustc_builtins-5c8a87f356eb958c.rlib(rustc_builtins-5c8a87f356eb958c.0.o): In function `__udivmodsi4':
rustc_builtins.cgu-0.rs:(.text.__udivmodsi4+0xb4): undefined reference to `core::panicking::panic::h6d47421a09de1633'
/tmp/rustc.n3iL9DJW0uUT/librustc_builtins-5c8a87f356eb958c.rlib(rustc_builtins-5c8a87f356eb958c.0.o): In function `__udivmoddi4':
rustc_builtins.cgu-0.rs:(.text.__udivmoddi4+0x304): undefined reference to `core::panicking::panic::h6d47421a09de1633'
rustc_builtins.cgu-0.rs:(.text.__udivmoddi4+0x30c): undefined reference to `core::panicking::panic::h6d47421a09de1633'
rustc_builtins.cgu-0.rs:(.text.__udivmoddi4+0x314): undefined reference to `core::panicking::panic::h6d47421a09de1633'
/tmp/rustc.n3iL9DJW0uUT/librustc_builtins-5c8a87f356eb958c.rlib(rustc_builtins-5c8a87f356eb958c.0.o): In function `__aeabi_uidiv':
rustc_builtins.cgu-0.rs:(.text.__aeabi_uidiv+0xa0): undefined reference to `core::panicking::panic::h6d47421a09de1633'
/tmp/rustc.n3iL9DJW0uUT/librustc_builtins-5c8a87f356eb958c.rlib(rustc_builtins-5c8a87f356eb958c.0.o):rustc_builtins.cgu-0.rs:(.text.__divsi3+0x40): more undefined references to `core::panicking::panic::h6d47421a09de1633' follow
collect2: error: ld returned 1 exit status

error: aborting due to previous error
```

However, the error only occurs if `lto = true` is set in Cargo.toml. If LTO is disabled, the result is a successful build.

### PR DESCRIPTION
**Pull request by japaric**
closes #79

### CODE REVIEW

**Comment by japaric** [comment]
r? @alexcrichton 

@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 88862e4 with merge 88862e4...

**Comment by alexcrichton** [comment]
r+ from me

Want to add this to the CI though to ensure there are no references to the panic symbol?

**Comment by japaric** [comment]
@alexcrichton sure thing

**Comment by japaric** [comment]
Hmm, so, with these changes, `librustc_builtins.rlib` still has undefined references to `panic` when built in release mode though I don't see any branch/jump instructions to it in the disassembly.

But, without these changes, if I build `intrinsics.rs` in release with LTO enabled then I can reproduce the linker errors seen in #79. And the errors disappear when I apply the changes in this PR. Would that suffice as a test?

**Comment by japaric** [comment]
> And the errors disappear when I apply the changes in this PR

Actually just the changes in this PR are not enough to remove all the panics. I'm going to homu try this so you can see the errors.

**Comment by japaric** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 7c3b64f with merge 7c3b64f...

**Comment by homunkulus** [comment]
:broken_heart: Test failed - [status-travis](https://travis-ci.org/japaric/rustc-builtins/builds/164210078)

**Comment by alexcrichton** [comment]
Panics are also inserted for `/` and `%` operations by default to check if the rhs would cause a fault (e.g. 0 or overflow w/ INT_MIN). Maybe that's happening here?

**Comment by japaric** [comment]
Hmm, does that means that we have to create wrappers around primitives that `intrinsics::abort` on bad inputs (division by zero) or e.g. use `intrinsics::unchecked_div` on good inputs?

**Comment by alexcrichton** [comment]
Ideally `unchecked_div` I think because otherwise this would rely on optimizations to remove references to the symbols.

**Comment by japaric** [comment]
@homunkulus try

**Comment by japaric** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit ee24d69 with merge ee24d69...

**Comment by homunkulus** [comment]
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/japaric/rustc-builtins/build/1.0.279), [status-travis](https://travis-ci.org/japaric/rustc-builtins/builds/165409074)
State: approved= try=True

**Comment by japaric** [comment]
@alexcrichton I added a few `intrinsics::abort` here and there where division by zero may occur and the linker errors are gone ...

**Comment by alexcrichton** [comment]
Hm yeah that works for release mode, but for debug mode the symbols would stick around?

**Comment by japaric** [comment]
I changed all the division and modulo operations with macros that abort on bad inputs. This makes the LTO work with the dev profile iff the debug-assertions are turned off ...

If we want LTO to work with debug assertions on, we'll have to change all the operations that may overflow with an abort on overflow version

**Comment by japaric** [comment]
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 34cd40f with merge 5766097...

**Comment by homunkulus** [comment]
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/japaric/rustc-builtins/build/1.0.285), [status-travis](https://travis-ci.org/japaric/rustc-builtins/builds/165848285)
State: approved= try=True

File: Cargo.toml [hunk 1]
```diff
@@ -25,4 +25,7 @@ compiler-rt = { path = "compiler-rt" }
 c = []
 weak = ["rlibc/weak"]
 
+[profile.dev]
+debug-assertions = false
```
Comment (alexcrichton) [inline Cargo.toml]: It seems desirable to leave this on, right? It's caught bugs in overflows and such?

File: src/lib.rs [hunk 2]
```diff
@@ -13,6 +13,66 @@
...
+}
+
+macro_rules! urem {
+    ($a:expr, $b:expr) => {
+        unsafe {
+            let a = $a;
+            let b = $b;
+
+            if b == 0 {
+                intrinsics::abort()
```
Comment (alexcrichton) [inline src/lib.rs]: You may want to make these `::core::intrinsics::...` to avoid imports in the other modules.

**Comment by alexcrichton** [comment]
> If we want LTO to work with debug assertions on, we'll have to change all the operations that may overflow with an abort on overflow version

What's the error message if debug assertions are turned on?

**Comment by japaric** [comment]
> What's the error message if debug assertions are turned on?

Building `src/bin/intrinsics.rs` for `arm-unknown-linux-gnueabi` using the dev profile (+debug-assertions) + LTO:

```
error: linking with `arm-linux-gnueabi-gcc` failed: exit code: 1
  |
  = note: "arm-linux-gnueabi-gcc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-L" "/rust/lib/rustlib/arm-unknown-linux-gnueabi/lib" "/target/arm-unknown-linux-gnueabi/debug/intrinsics.0.o" "-o" "/target/arm-unknown-linux-gnueabi/debug/intrinsics" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/target/arm-unknown-linux-gnueabi/debug/deps" "-L" "/target/arm-unknown-linux-gnueabi/debug/build/rustc_builtins-3d8222040fb6ac11/out" "-L" "/rust/lib/rustlib/arm-unknown-linux-gnueabi/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/tmp/rustc.1PHTT1MGYLXp/librustc_builtins.rlib" "-l" "c" "-l" "m" "-l" "rt" "-l" "util"
  = note: /tmp/rustc.1PHTT1MGYLXp/librustc_builtins.rlib(rustc_builtins.0.o): In function `rustc_builtins::int::udiv::__udivmoddi4':
/checkout/src/int/udiv.rs:210: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/checkout/src/int/udiv.rs:217: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/checkout/src/int/udiv.rs:220: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/checkout/src/int/udiv.rs:220: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/checkout/src/int/udiv.rs:221: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/tmp/rustc.1PHTT1MGYLXp/librustc_builtins.rlib(rustc_builtins.0.o):/checkout/src/int/udiv.rs:198: more undefined references to `core::panicking::panic::h7215d66c75dd9558' follow
collect2: error: ld returned 1 exit status


error: aborting due to previous error

error: Could not compile `rustc_builtins`.
```

**Comment by japaric** [comment]
> /checkout/src/int/udiv.rs:210

is: `if sr > u32::bits() - 1 {`

> /checkout/src/int/udiv.rs:217

is `sr += 1;`

and so on.

**Comment by alexcrichton** [comment]
Oh those just need to be wrapping ops, right? Or explicitly `.checked_foo` with an abort?

**Comment by japaric** [comment]
> It seems desirable to leave this on, right? It's caught bugs in overflows and such?

I don't know if these have helped catch bugs but we do have seen debug assertions related to overflows while testing the intrinsics.

> Oh those just need to be wrapping ops, right? Or explicitly .checked_foo with an abort?

I prefer checked + abort. Although not as nice as panics, that way still tells you there's overflow going on so you can look at the offending code and decide if overflowing makes sense. OTOH, silent overflows sound like a bad default.

**Comment by alexcrichton** [comment]
Yeah I agree that we probably want to catch overflows as much as possible. Want to try out the checked + abort strategy?

**Comment by japaric** [comment]
Hmm, would it make sense to change *all* the arithmetic operations to checked+abort?

**Comment by japaric** [comment]
Hmm, it seems like the `overflowing_`, `checked_` variants have `panic!`s inside them?

I changed the contents of the macros to `.checked_*().unwrap_or_else(|| abort)` and I get:

```
error: linking with `arm-linux-gnueabi-gcc` failed: exit code: 1
  |
  = note: "arm-linux-gnueabi-gcc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-L" "/rust/lib/rustlib/arm-unknown-linux-gnueabi/lib" "/target/arm-unknown-linux-gnueabi/debug/intrinsics.0.o" "-o" "/target/arm-unknown-linux-gnueabi/debug/intrinsics" "-Wl,--gc-sections" "-pie" "-nodefaultlibs" "-L" "/target/arm-unknown-linux-gnueabi/debug/deps" "-L" "/target/arm-unknown-linux-gnueabi/debug/build/rustc_builtins-3d8222040fb6ac11/out" "-L" "/rust/lib/rustlib/arm-unknown-linux-gnueabi/lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "/tmp/rustc.ZoLec8ROwmlU/librustc_builtins.rlib" "-l" "c" "-l" "m" "-l" "rt" "-l" "util"
  = note: /tmp/rustc.ZoLec8ROwmlU/librustc_builtins.rlib(rustc_builtins.0.o): In function `core::num::{{impl}}::overflowing_div':
/buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-host-linux/build/src/libcore/num/mod.rs:1013: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-host-linux/build/src/libcore/num/mod.rs:1013: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/tmp/rustc.ZoLec8ROwmlU/librustc_builtins.rlib(rustc_builtins.0.o): In function `core::num::{{impl}}::checked_div':
/buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-host-linux/build/src/libcore/num/mod.rs:1691: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/tmp/rustc.ZoLec8ROwmlU/librustc_builtins.rlib(rustc_builtins.0.o): In function `core::num::{{impl}}::checked_rem':
/buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-host-linux/build/src/libcore/num/mod.rs:1712: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/tmp/rustc.ZoLec8ROwmlU/librustc_builtins.rlib(rustc_builtins.0.o): In function `core::num::{{impl}}::checked_div':
/buildslave/rust-buildbot/slave/nightly-dist-rustc-cross-host-linux/build/src/libcore/num/mod.rs:1691: undefined reference to `core::panicking::panic::h7215d66c75dd9558'
/tmp/rustc.ZoLec8ROwmlU/librustc_builtins.rlib(rustc_builtins.0.o):/checkout/src/int/udiv.rs:210: more undefined references to `core::panicking::panic::h7215d66c75dd9558' follow
collect2: error: ld returned 1 exit status


error: aborting due to previous error

error: Could not compile `rustc_builtins`.
```

**Comment by alexcrichton** [comment]
> Hmm, would it make sense to change all the arithmetic operations to checked+abort?

I think to get rid of references to the panic symbol, yeah.

> Hmm, it seems like the overflowing_, checked_ variants have panic!s inside them?

Oh gah it looks like this is a bug in the standard library, they still use the standard `/` operators, which in turn have panics. I think we'd have to upstream a patch to use the intrinsics without panics in these cases.

**Comment by japaric** [comment]
> Oh gah it looks like this is a bug in the standard library, they still use the standard / operators, which in turn have panics.

The implementation of `overflowing_div`, which is called by `checked_div`, looks like this:

``` rust
        pub fn overflowing_div(self, rhs: Self) -> (Self, bool) {
            if self == Self::min_value() && rhs == -1 {
                (self, true)
            } else {
                unsafe {
                    (self / rhs, false)
                }
            }
        }
```

but we can't change the `self / rhs` to `unchecked_div(self, rhs)` because `rhs = 0` would be UB and this function doesn't protect against division by zero. `:-/`

**Comment by alexcrichton** [comment]
Hm ok, perhaps a suite of checked ops could be added to this crate?

**Comment by alexcrichton** [comment]
Note that this should likely be fixed upstream as well

**Comment by japaric** [comment]
Ugh, I've been using macros to do the checked.unwrap_or(abort) thing but pretty much every single operation has to be replaced by that macro, which in results in very verbose code. I think that we instead should use newtypes with abort on bad inputs/overflow semantics to avoid obfuscating the code. But that's going to require a bunch of boilerplate instead.

Can we punt fixing debug (-debug-assertions) + LTO and land this PR as it's (modulo nits)?

**Comment by alexcrichton** [comment]
Hm so disabling debug assertions here won't actually work for everyone else. That is, the top-level project decides this, not the crate itself. Put another way this won't fix #79 (the intention, right?) because the `debug-assertions = false` will be ignored.

I'd probably recommend keeping the same form of style here and avoiding macros. It's already kinda hard to follow with macros generating functions :(.

Perhaps some new types could be added whose operators (like `+` and such) are overloaded to abort on overflows?

**Comment by japaric** [comment]
> Hm so disabling debug assertions here won't actually work for everyone else.

Oh, yes. I'm aware. That what actually just for testing this on the debug-assertion-less scenario.

> Put another way this won't fix #79 (the intention, right?)

No, this won't totally fix #79. But it does fix the LTO + release case which is, I think, the combination that most people care about. @FenrirWolf, are you mostly interested in profile.release + LTO? Or is profile.dev + LTO also important to you?

> Perhaps some new types could be added whose operators (like + and such) are overloaded to abort on overflows?

Certainly that's a better fix and I've also suggested it above. What I wanted to do is fix profile.release + LTO now and then implement the newtype approach to fix the general case.

**Comment by FenrirWolf** [comment]
@japaric I haven't been using profile.dev + LTO since that combination was producing an LLVM error for a long time. Though now that you've reminded me, I just checked to see if that error still happens on the latest nightly, and it doesn't. So that's cool.

Either way, I've mainly been using profile.release + LTO. I can technically go without that too if need be.

**Comment by alexcrichton** [comment]
@japaric ok if this is targeted at fixing the release + LTO case, sounds good to me! Let's go ahead and land to then fixup the debug case later

**Comment by japaric** [comment]
@alexcrichton r?
@homunkulus try

**Comment by homunkulus** [comment]
:hourglass: Trying commit 96ddd4e with merge 96ddd4e...

**Comment by alexcrichton** [comment]
r+

**Comment by homunkulus** [comment]
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/japaric/rustc-builtins/build/1.0.289), [status-travis](https://travis-ci.org/japaric/rustc-builtins/builds/165963160)
State: approved= try=True
